### PR TITLE
Create runtime-transaction crate to host transaction type for runtime

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem

Add a transaction type for runtime that contains encapsulated internal metadata, which can be accessed/updated as sanitized transaction goes through execution pipeline.

#### Summary of Changes
- create a crate for runtime-transaction type to reduce dependencies
- add `RuntimeTransaction` struct that can be built from `sdk::SanitizedVersionedTransaction`
- add `TransactionMeta` currently only has `is_simple_vote_tx` flag to match `sdk::SanitizedTransaction`. Will be expanded later (see comment)
- add helper function `is_simple_vote_transaction(...)` 

Fixes #

### CODE REVIEW

File: runtime-transaction/src/runtime_transaction.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            todo!();
+        }
+        Ok(())
+    }
+
+    // Getter
+    pub fn get_message_hash(&self) -> &Hash {
+        if let RuntimeMessage::Uninitialized = self.message {
+            panic!("Runtime transaction is uninitialized");
+        }
```
Comment (taozhu-chicago) [inline runtime-transaction/src/runtime_transaction.rs:130]: panic or return Option when getter is called in wrong state? I pick panicking to fail hard, but mostly to keep interface simple.

File: runtime-transaction/src/runtime_transaction.rs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+///
+/// New Error:
+/// If Getter of account-based metadata is called while in Initial state,
+/// a new Error shall return to indicate RuntimeTransaction is not
+/// furnished with necessary data yet.
+///
+#[derive(Debug, Default, Clone, Eq, PartialEq)]
+pub enum RuntimeMessage {
+    #[default]
+    Uninitialized,
```
Comment (taozhu-chicago) [inline runtime-transaction/src/runtime_transaction.rs:54]: this is unnecessary, strictly speaking, need it for `std::mem::take()`

File: runtime-transaction/src/runtime_transaction.rs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    StaticallyLoaded(SanitizedVersionedMessage),
+    DynamicallyLoaded(SanitizedMessage),
```
Comment (taozhu-chicago) [inline runtime-transaction/src/runtime_transaction.rs:56]: enum variants VS individual structs, can't decide. 

with enum, i have one `RuntimeTransaction` that can be in 3 states; But call site needs to `match ...` all the time?

Or can have three separate structs: `UninitializedRuntimeTransaction`,`StaticallyLoadedRuntimeTransaction` and `DynamicallyLoadedRuntimeTransaction`. Awful names, but strong typed, so cleaner. One question is if a function takes `StaticallyLoadedRuntimeTransaction`, but you have `DynamicallyLoadedRuntimeTransaction`, how do you downcast?

Comment (apfitzge) [reply runtime-transaction/src/runtime_transaction.rs:56]: Dealing with the "downcasting" probably has 2 approaches:

1. w/ separate structs - each struct implements some trait to access some info
2. w/ enums - top-level fn, match & access

File: runtime-transaction/src/transaction_meta.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+//!
+
+use {
+    solana_sdk::hash::Hash,
+};
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct TransactionMeta {
+    pub message_hash: Hash,
+    pub is_simple_vote_tx: bool,
```
Comment (apfitzge) [inline runtime-transaction/src/transaction_meta.rs:15]: these feel like they should be `Option<T>` so we know if they've been populated or not

**Comment by apfitzge** [comment]
Some thoughts:

## enum vs structs

- Separate structs allow us know @ compile time what the state of the message is
- Enums give us flexibility on when we load ALTs

I think the big question here is...do we need that flexibility? ...maybe?
It's probably inefficient for the scheduling thread to load addresses **during** scheduling.

I've played with separate threads receiving & sanitizing outside the scheduling loop, which would probably be better off w/ separate structs so that we know we always have the addresses from ALTs. Additionally we could still be doing checks and discard before loading addresses here.

Have kinda been debating whether we want these sorts of meta-tagged structs, rather than just having some final state:

```rust
struct RuntimeTransaction {
    signatures: Vec<Signature>,
    message: SanitizedMessage,
    meta: TransactionMeta,
}
```

and up to the point of transition into a `RuntimeTransaction` we just keep pairs as an internal book-keeping: `(SanitizedVersionedTransaction, TransactionMeta)`

We can do basic checking on the `SanitizedVersionedTransaction`, populate (static) meta fields, and potentially discard the tx and meta if we hit some cases.
Then move it into a `RuntimeTransaction`, loading addresses and populating dynamic meta fields (in either order?)
This suggestion is not really different than the separate struct approach.
Kinda leaning towards separate structs, or just a single final struct.

**Comment by taozhu-chicago** [comment]
I also leaning to #33471 compile time binding with separate structs (just needs better names).

Transaction_meta is open-ended, can add more stuff into it as new logic are introduced. (we have freedom now since it's in own crate facing runtime). Few traits are defined for different states, separate structs can implement proper traits to access meta.

Closing this PR in fav of #33471 , move discussino there.

**Pull request closed without merging**